File: code/scenes/src/main/java/org/fbme/scenes/controllers/diagram/ConnectionCursor.java
package org.fbme.scenes.controllers.diagram;

import org.jetbrains.annotations.NotNull;

public class ConnectionCursor<ConnT, CursorT> {

    @NotNull
    private final ConnT myConnection;

    @NotNull
    private final CursorT myCursor;

    public ConnectionCursor(@NotNull ConnT connection, @NotNull CursorT cursor) {
        myConnection = connection;
        myCursor = cursor;
    }

    public ConnT getConnection() {
        return myConnection;
    }

    public CursorT getCursor() {
        return myCursor;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ConnectionCursor)) {
            return false;
        }

        ConnectionCursor<?, ?> that = (ConnectionCursor<?, ?>) o;
        return myConnection.equals(that.myConnection) && myCursor.equals(that.myCursor);
    }

    @Override
    public int hashCode() {
        int result = 0;
        result = 31 * result + myConnection.hashCode();
        result = 31 * result + myCursor.hashCode();
        return result;
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/diagram/ConnectionCursor.kt
package org.fbme.scenes.controllers.diagram

class ConnectionCursor<ConnT, CursorT>(val connection: ConnT, val cursor: CursorT) {
    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is ConnectionCursor<*, *>) {
            return false
        }
        return connection == other.connection && cursor == other.cursor
    }

    override fun hashCode(): Int {
        var result = 0
        result = 31 * result + connection.hashCode()
        result = 31 * result + cursor.hashCode()
        return result
    }
}